### PR TITLE
Update configuration & add types for API responses

### PR DESCRIPTION
**Pull request by andrewatwood**
API responses now support generics that allow you to specify the shape of the expected response and prevent the use of type guards to ensure specific properties exist on a given API response

- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

- **What is the current behavior?** (You can also link to an open issue here)
Lots of `:?` in the API response interface, which leads to a lot of `undefined` checks that don't need to be there when we know what shape the response really has.

- **What is the new behavior (if this is a feature change)?**
API responses now support generics that allow you to specify the shape of the expected response and prevent the use of type guards to ensure specific properties exist on a given API response. Also some other smaller housekeeping things such as throwing errors for missing config params, and a helper error class to help give useful error messages.

### CODE REVIEW

File: src/lib/model/index.ts [hunk 1]
```diff
@@ -1 +1,7 @@
+export * from './account-vo';
+export * from './archive-vo';
+export * from './folder-vo';
+export * from './record-vo';
 export * from './request-vo';
+export * from './simple-vo';
+export * from './share-by-url-vo';
```
Comment (jasonaowen) [inline src/lib/model/index.ts:7]: Alphabetize:

```suggestion
export * from './share-by-url-vo';
export * from './simple-vo';
```

(We should really set up stricter linter rules; it's such a silly thing to comment on in a code review!)

File: src/lib/client.ts [hunk 2]
```diff
@@ -1,30 +1,48 @@
...
+      throw new PermSdkError('Missing sessionToken in config');
+    }
+
+    if (!mfaToken) {
+      throw new PermSdkError('Missing mfaToken in config');
+    }
+
+    if (!archiveNbr) {
+      throw new PermSdkError('Missing archiveNbr in config');
+    }
```
Comment (jasonaowen) [inline src/lib/client.ts:35]: These are a good first step towards input validation!

I'll note that we're still not validating the type of the input, only its presence - we could call this from JavaScript with `new Permanent({apiKey: 1, ...etc})` and these checks would pass. But still, this is good progress towards #7 - thanks!

Also, specifically for this parameter, is it safe to say that `0` is never a valid `archiveNbr`?

**Review by jasonaowen — APPROVED** [review]
This is laying a good foundation! Thanks for carving it off from #8, @andrewatwood .

A few trivial suggestions below, so marking as approved so you can fix and merge without blocking on me. :slightly_smiling_face: 

I was able to build and run this locally, and verify that omitting parameters in the constructor caused errors:

```
$ npm run build
$ nodejs
> const permsdk = require('./build/main/');
undefined
> new permsdk.Permanent({archiveNbr: 0, sessionToken: 1, mfaToken: 2, apiKey: 3});
Uncaught PermSdkError: @permanentorg/node-sdk - Missing archiveNbr in config
    at new Permanent (./build/main/lib/client.js:17:19)
```

I think the various type definitions are good but not very testable from a user point of view, so I didn't do anything there - CI passed, so that's good enough for me!

File: src/lib/model/request-vo.ts [hunk 3]
```diff
@@ -1,15 +1,31 @@
-import { SimpleVO } from './simple-vo';
+import { RecordVO } from './record-vo';
 
-export interface PermanentApiData {
-  FolderVO?: unknown;
-  RecordVO?: unknown;
-  ArchiveVO?: unknown;
-  AccountVO?: unknown;
+import { AccountVO, ArchiveVO, FolderVO, ShareByUrlVO, SimpleVO } from '.';
```
[the remark this answers is not in the document]
Comment (andrewatwood) [reply src/lib/model/request-vo.ts:3]: Oh heh whoops, yeah that stragger is ugly. I like that convention, will go with the latter option to follow it!

File: src/lib/client.ts (hunk 2 above)
Comment (andrewatwood) [reply src/lib/client.ts:35]: Safe to say, yes! Valid `archiveNbr` will always be a string.

#7 will be great, coming soon!

File: src/lib/model/request-vo.ts [hunk 4]
```diff
@@ -1,15 +1,31 @@
...
+  RecordVO?: RecordVO;
+  ArchiveVO?: ArchiveVO;
+  AccountVO?: AccountVO;
   SimpleVO?: SimpleVO;
+  Shareby_urlVO?: ShareByUrlVO;
+}
+
+export type PermanentApiResponseData<
+  T extends keyof PermanentApiResponseDataBase
+> = Required<Pick<PermanentApiResponseDataBase, T>>;
```
[the remark this answers is not in the document]
Comment (andrewatwood) [reply src/lib/model/request-vo.ts:19]: This is by far the most "in the weeds" thing I've done with generics, was thrilled that it worked relatively painlessly.

**Pull request merged**